File: frontend/lib/models/shopping_list.dart
class ShoppingListItem {
  final String id;
  final String? productId;
  final String productName;
  final String? brand;
  final String departmentName;
  final int departmentSortOrder;
  final double requiredQuantity;
  final String unit;
  final double? estimatedPrice;
  bool isChecked;
  final String? substitutedForName;

  ShoppingListItem({
    required this.id,
    this.productId,
    required this.productName,
    this.brand,
    required this.departmentName,
    required this.departmentSortOrder,
    required this.requiredQuantity,
    required this.unit,
    this.estimatedPrice,
    required this.isChecked,
    this.substitutedForName,
  });

  factory ShoppingListItem.fromJson(Map<String, dynamic> json) {
    return ShoppingListItem(
      id: json['id'] as String,
      productId: json['product_id'] as String?,
      productName: json['product_name'] as String,
      brand: json['brand'] as String?,
      departmentName: json['department_name'] as String? ?? 'Inne',
      departmentSortOrder: json['department_sort_order'] as int? ?? 99,
      requiredQuantity: (json['required_quantity'] as num? ?? 0.0).toDouble(),
      unit: json['unit'] as String? ?? 'szt',
      estimatedPrice: json['estimated_price'] != null
          ? (json['estimated_price'] as num).toDouble()
          : null,
      isChecked: json['is_checked'] as bool? ?? false,
      substitutedForName: json['substituted_for_name'] as String?,
    );
  }
}

class ShoppingList {
  final String id;
  final String mealPlanId;
  final String storeId;
  final String storeName;
  final String status;
  final double totalEstimatedPrice;
  final Map<String, List<ShoppingListItem>> itemsByDepartment;
  final DateTime createdAt;

  ShoppingList({
    required this.id,
    required this.mealPlanId,
    required this.storeId,
    required this.storeName,
    required this.status,
    required this.totalEstimatedPrice,
    required this.itemsByDepartment,
    required this.createdAt,
  });

  factory ShoppingList.fromJson(Map<String, dynamic> json) {
    final rawGrouped = json['items_by_department'] as Map<String, dynamic>? ?? {};
    final parsedGrouped = <String, List<ShoppingListItem>>{};

    rawGrouped.forEach((key, value) {
      if (value is List) {
        parsedGrouped[key] = value
            .map((item) => ShoppingListItem.fromJson(item as Map<String, dynamic>))
            .toList();
      }
    });

    return ShoppingList(
      id: json['id'] as String,
      mealPlanId: json['meal_plan_id'] as String,
      storeId: json['store_id'] as String,
      storeName: json['store_name'] as String? ?? '',
      status: json['status'] as String? ?? 'pending',
      totalEstimatedPrice: (json['total_estimated_price'] as num? ?? 0.0).toDouble(),
      itemsByDepartment: parsedGrouped,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  int get totalItems {
    return itemsByDepartment.values.fold(0, (sum, list) => sum + list.length);
  }

  int get checkedItems {
    return itemsByDepartment.values.fold(
      0,
      (sum, list) => sum + list.where((item) => item.isChecked).length,
    );
  }

  double get progress {
    final total = totalItems;
    if (total == 0) return 0.0;
    return checkedItems / total;
  }
}

File: backend/app/schemas/shopping_list.py
"""Schematy Pydantic v2 dla list zakupów."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, model_validator


class ShoppingListItemResponse(BaseModel):
    """Odpowiedź API — pozycja na liście zakupów."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    product_id: uuid.UUID | None = None
    product_name: str
    brand: str | None = None
    department_name: str | None = None
    department_sort_order: int | None = None
    required_quantity: float
    unit: str
    estimated_price: float | None = None
    is_checked: bool
    substituted_for_name: str | None = None

    @model_validator(mode="before")
    @classmethod
    def get_related_fields(cls, data):
        if hasattr(data, "store_product") and data.store_product is not None:
            prod = data.store_product.product
            dept = data.department
            return {
                "id": data.id,
                "product_id": prod.id if prod else None,
                "product_name": prod.name if prod else "Nieznany produkt",
                "brand": prod.brand if prod else None,
                "department_name": dept.name if dept else "Inne",
                "department_sort_order": dept.sort_order if dept else 99,
                "required_quantity": data.required_quantity,
                "unit": data.unit,
                "estimated_price": data.estimated_price,
                "is_checked": data.is_checked,
                "substituted_for_name": (
                    data.substituted_for_product.name
                    if getattr(data, "substituted_for_product", None) is not None
                    else None
                ),
            }
        return data


class ShoppingListResponse(BaseModel):
    """Odpowiedź API — pełna lista zakupów pogrupowana wg działów."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    meal_plan_id: uuid.UUID
    store_id: uuid.UUID
    store_name: str
    status: str
    total_estimated_price: float | None = None
    items_by_department: dict[str, list[ShoppingListItemResponse]] = {}
    created_at: datetime

    @model_validator(mode="before")
    @classmethod
    def get_related_fields(cls, data):
        if hasattr(data, "store") and data.store is not None:
            items_by_dept = {}
            total_price = 0.0
            for item in (data.items or []):
                dept_name = item.department.name if item.department else "Inne"
                if dept_name not in items_by_dept:
                    items_by_dept[dept_name] = []
                items_by_dept[dept_name].append(item)
                if item.estimated_price:
                    total_price += float(item.estimated_price)
            
            return {
                # UWAGA: zwracamy meal_plan_id jako "id", ponieważ wszystkie endpointy
                # listy zakupów (/shopping-lists/{list_id}/...) identyfikują listę po ID
                # planu posiłków (relacja 1:1), a nie po własnym kluczu głównym ShoppingList.
                # Zwracanie tu data.id (PK listy) powodowało, że każda kolejna akcja
                # (zaznaczanie produktu, podmiana zamiennika, podsumowanie) kończyła się 404.
                "id": data.meal_plan_id,
                "meal_plan_id": data.meal_plan_id,
                "store_id": data.store_id,
                "store_name": data.store.name,
                "status": data.status,
                "total_estimated_price": round(total_price, 2),
                "items_by_department": items_by_dept,
                "created_at": data.created_at,
            }
        return data
